lib/Sema/SemaBounds.cpp

```diff
@@ -2058,6 +2060,126 @@ namespace {
       return BaseRange::Kind::ConstantSized;
     }
 
+    // In this function, the goal is to compare two expressions: `Base1 + Offset1` and `Base2 + Offset2`
```

Comment:
I think a more general approach would be to try to put each arithmetic expression into a standard form involving integer arithmetic.  If that is possible, you can then compare the standard forms.   The standard form would be base + (const op e), where op is either signed multiplication or unsigned multiplication, and base, const, and e are all interpreted as integers.   You could have variables that represent each of the components of the standard form.

Given `array_ptr<int>` p:
- For `((char *) p) + e1`, the standard form is `base = ((char *) p, const = 1, e = e1.`  op depends on the signedness of e1.
- For `p + e2`. the standard form is `base = p, const = sizeof(int), e = e2`.  op depends on the signedness of `e2`.
- For `((unsigned char *) p) + e1`, the standard from is `base = (unsigned char *) p, const = 1, e = e1`.  op depends on the signedness of `e1`.

To check that two expressions to produce identical results, we would check that the multiplication operators have the same signedness.   At this time, I don't think we should try to handle the case where one multiplication operator is signed and the other is unsigned (we would need additional facts to prove that the two operations produce the same value: that the signed integer is >= 0 and that the multiplication does not overflow).